CIPs/cip-0032.md

```diff
@@ -13,81 +13,105 @@ license: Apache 2.0
+
+- `TargetCompletionRate` - The target completion rate for Attestation Services, taking into account standard user drop-off and other factors outside of the AS control.
+- `AttestationServiceRewardPercentage` - The percentage of max rewards that are allocated to running an Attestation Service. Initially `.20`.
+- `MinGlobalAttestationCompletionRate` - The minimum completion rate across all validators for rewards to be reduced for registered Attestation Services.
+- `GlobalAttestationFailureFallbackReward` - The rate that Attestation Service operators will be rewarded when the global completion rate drops below `MinGlobalAttestationCompletionRate`.
 
-- `MinAttestationSlashablePeriod` is the number of consecutive blocks in which an elected Attestation Service is required to complete at least 1 attestation.
-- `AttestationSlasherPenalty` is the penalty amount that is slashed in a slashing event.
-- `AttestationSlasherReward` is the reward amount that a slasher is granted.
-- `MinAttestationSlashablePeriod`, `AttestationSlasherPenalty`, and `AttestationSlasherReward` are easily configurable via governance proposals.
+- `AttestationMinPeriod` - The period which the issuer must complete at least one successful attestation or potentially face deregistration.
+- `AttestationMinRequested` - The minimum number of attestations requested from the issuer to be eligible for removal.
+- `MinNumberOfAttestationIssuers` - If deregistering an Attestation Service would drop the count below this number, then prevent removal.
 
-This proposal introduces a new `AttestationSlasher` contract which allows slashing of stake if a completed attestation hasn't been observed in the past `MinAttestationSlashablePeriod` blocks. The `Attestations` contract will be upgraded to register the selection and completion events in the new `AttestationSlasher` contract. The `AttestationSlasher` will contain the following state to track the latest block number that an attestation was requested/completed on:
+The `Attestations` contract will be upgraded to register the request and completion events. It will be read by the new `Validators` contract. The `Attestations` contract will contain the following state to track the request and completion counts per epoch:
 
 ```js
-// Tracks the last time an attestation was requested from an issuer (AS node)
-mapping(address => uint256) public lastAttestationRequested;
+// Tracks per issuer completion rate
+// These values are reset per issuer from Validators.distributeEpochPaymentsFromSigner()
+mapping(address => uint256) public attestationRequestCount;
+mapping(address => uint256) public attestationCompletionCount;
+
+// Track global completion rate
+// currEpochIndex = getEpochNumberOfBlock(block.number) % 2
```

Comment:
Why use `% 2` when a full 8 bits are available? We could change this to `% 256`. It's probably the same either way, because the only case where this would matter is if not requests are seen for a full epoch, which is unlikely, but it's also not impossible. (Also see comment below as there may actually be no advantage to using `uint8` here anyway.)

Reply:
Good point!